gobblin-metrics/src/main/java/gobblin/metrics/KafkaAvroReporter.java

```diff
@@ -0,0 +1,161 @@
+      " ]\n"+
+      "}";
+
+  Schema _schema;
+  private GenericDatumWriter<GenericRecord> w;
+  private Encoder e;
+  private ByteArrayOutputStream out;
+  private int exceptionCount;
+
+  protected KafkaAvroReporter(MetricRegistry registry, String name, MetricFilter filter, TimeUnit rateUnit,
+      TimeUnit durationUnit, String brokers, String topic, String host, String env, Set<String> tags) {
+    super(registry, name, filter, rateUnit, durationUnit, brokers, topic, host, env, tags);
+
+    Schema.Parser parser = new Schema.Parser();
+    _schema = parser.parse(SchemaString);
+
+    exceptionCount = 0;
+
+    out = new ByteArrayOutputStream();
+    e = EncoderFactory.get().binaryEncoder(out, null);
+    w = new GenericDatumWriter<GenericRecord>(_schema);
+
+  }
+
+  /**
+   * Serializes a metric key-value pair into an avro object
+   * Calls _serializeValue, which is synchronized because avro serializer is not thread-safe
+   *
+   * @param name name of the metric
+   * @param value value of the metric to report
+   * @param path additional suffixes to further identify the meaning of the reported value
+   * @return
+   */
+  @Override
+  protected byte[] serializeValue(String name, Object value, String... path) {
+    return _serializeValue(name, value, path);
+  }
+
+  protected synchronized byte[] _serializeValue(String name, Object value, String... path) {
+
+    GenericRecord record = new GenericData.Record(_schema);
+
+    record.put("name", makeName(name, path));
+    record.put("value", value);
+    record.put("hostname", _host);
+    record.put("environment", _env);
+    record.put("tags", _tags);
+
+    try {
+      out.reset();
+      w.write(record, e);
+      e.flush();
+      return out.toByteArray();
+    } catch(IOException e) {
+      exceptionCount++;
+      LOGGER.trace("Could not serialize Avro record for Kafka Metrics. Exception: %s", e.getMessage());
+      if (exceptionCount % 1000 == 0) {
+        LOGGER.warn("Could not serialize Avro record for Kafka Metrics. Exception: %s", e.getMessage());
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Returns a new {@link gobblin.metrics.KafkaAvroReporter.Builder} for {@link gobblin.metrics.KafkaAvroReporter}
+   *
+   * @param registry the registry to report
+   * @return KafkaAvroReporter builder
+   */
+  public static Builder forRegistry(MetricRegistry registry) {
+    return new BuilderImpl(registry);
+  }
+
+  private static class BuilderImpl extends Builder<BuilderImpl> {
+    public BuilderImpl(MetricRegistry registry) {
+      super(registry);
+    }
+
+    @Override
+    protected BuilderImpl self() {
+      return this;
+    }
+  }
+
+  /**
+   * Builder for {@link gobblin.metrics.KafkaAvroReporter}
+   * Defaults to no filter, reporting rates in seconds and times in milliseconds
+   */
+  public static abstract class Builder<T extends Builder<T>> extends KafkaReporter.Builder<T> {
```

Comment:
This is a trick to allow extending the Builder in subclasses of KafkaAvroReporter. That way, subclasses don't have to reimplement all setters, etc. It is less important here than in KafkaReporter, but if anyone wants to extend it in the future, it will be useful.